### PR TITLE
Tp2000 1327  upgrade python 3.12

### PR DESCRIPTION
**Pull request by mattjamc**
# TP2000-1327 Upgrade Python 3.12

## Why

Upgrade Python to the latest version.

## What

Following on from the [PR to upgrade to 3.11](https://github.com/uktrade/tamato/pull/1228), this PR:

- Updates Python required to 3.12
- Upgrades coverage from 5.4 to 7.4 [Solved various bdd tests that began failing and reduced [test run time which had doubled](https://github.com/nedbat/coveragepy/issues/1665)]
- Upgrades lxml from 4.9.1 to 5.2.2 [Unable to build lxml version 4.9.1 - Error: Please make sure the libxml2 and libxslt development packages are installed]
- Following lmxl upgrade, makes bomb.xml file a bit longer to ensure it correctly raises the XMLSyntaxError
- Upgrades moto from 2.1.0 to 3.0.7 [Resolve AttributeError: module 'ssl' has no attribute 'wrap_socket']
- Following moto upgrade, accounts for BucketAlreadyOwnedByYou exception
- Removes redundant test_workbasket_clean_does_not_run_business_rules
- Upgrades pytest from 7.1.2 to 7.4.0 to get rid of [new deprecation warnings](https://github.com/pytest-dev/pytest/issues/10977)
- downgrades boto to stop this new error when sending a WB for packaging (botocore.exceptions.ClientError: An error occurred (PermanentRedirect) when calling the PutObject operation: The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.)
- downgrades elastic-apm to prevent the following error when downloading an envelope ("failed to validate span: error validating JSON: I[#] S[#] doesn't validate with \\\"span#\\\"\\n  I[#] S[#/allOf/3] allOf failed\\n    I[#/context/tags] S[#/allOf/3/properties/context/properties/tags/additionalProperties] additionalProperties \\\"aws.s3.bucket\\\" not allowed\)
- Upgrade pii-secret-check-hooks pre-commit

### CODE REVIEW

File: common/tests/test_files/bomb.xml [hunk 1]
```diff
@@ -3,5 +3,7 @@
 <!ENTITY b "&a;&a;&a;&a;&a;&a;&a;&a;">
 <!ENTITY c "&b;&b;&b;&b;&b;&b;&b;&b;">
 <!ENTITY d "&c;&c;&c;&c;&c;&c;&c;&c;">
+<!ENTITY e "&d;&d;&d;&d;&d;&d;&d;&d;">
+<!ENTITY f "&e;&e;&e;&e;&e;&e;&e;&e;">
 ]>
-<bomb>&c;</bomb>
\ No newline at end of file
+<bomb>&f;</bomb>
```
Comment (paulpepper-trade) [inline common/tests/test_files/bomb.xml:9]: Out of curiosity, why has the payload been made bigger and why the change from `&c;` to `&f;`?

Comment (mattjamc) [reply common/tests/test_files/bomb.xml:9]: Upgrading lxml made the test that uses this file start failing - like it no longer considered it an xml bomb so no longer errored. Maybe because it wasn't long enough so actually wasn't that much effort to parse.
